File: lma-desktop-capture-app-stack/source/windows/Engine/VideoSocket.cs
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;

namespace LMA;

/// <summary>
/// Second WebSocket, dedicated to the optional desktop-video stream:
///   1. connect wss://.../api/v1/ws  (same endpoint + query-param auth as the
///      audio socket — see the AUTH NOTE in TranscriberSocket.cs)
///   2. send one JSON text frame { callEvent: "START_VIDEO", callId, ... }
///   3. stream fragmented-MP4 segments as binary frames
///   4. send { callEvent: "END_VIDEO" } and close
///
/// Keeping video on its own socket means video bytes can never delay the
/// real-time audio PCM (no head-of-line blocking), and old servers — which
/// ignore the unknown START_VIDEO event and drop binary frames on a
/// session-less socket — degrade to exactly the pre-video behavior.
///
/// Reconnect semantics differ from audio: the fMP4 encoder session on our side
/// keeps running across a socket drop, so the bytes CONTINUE one stream. On
/// reconnect we re-send START_VIDEO with videoResume=true and flush everything
/// buffered — segments must never be dropped mid-stream (unlike audio, a gap
/// makes the remainder undecodable). If the outage outlasts the buffer cap we
/// give up on video for this call (audio is the priority) via OnOverflow.
///
/// Ported from macOS VideoSocket.swift.
/// </summary>
public sealed class VideoSocket
{
    private readonly Config _config;
    private ClientWebSocket? _ws;
    private CancellationTokenSource? _cts;
    private volatile bool _isOpen;
    private readonly object _sendLock = new();

    private volatile bool _intentionalClose;
    private int _reconnectAttempt;
    private const double MaxBackoff = 10;
    private readonly object _reconnectLock = new();
    private bool _reconnectScheduled;
    private bool _everOpened;
    private int _handshakeFailures;
    private const int MaxHandshakeFailures = 4;

    /// <summary>True after the first successful START_VIDEO: reconnects resume.</summary>
    private bool _startedOnce;

    /// <summary>
    /// ms between the audio stream's start and the first video frame; sent on
    /// START_VIDEO so the server can align video with the transcript timeline.
    /// </summary>
    public int VideoTimeOffsetMs;

    // Segments buffered while the socket is down. Generous cap (~32MB) because
    // dropping mid-stream corrupts the video.
    private readonly List<byte[]> _pending = new();
    private int _pendingBytes;
    private const int MaxPendingBytes = 32 * 1024 * 1024;
    private volatile bool _overflowed;

    /// <summary>
    /// Fired once if the buffer cap is exceeded during an outage (or the video
    /// handshake keeps failing) — the caller should stop video capture for this
    /// call. Audio continues unaffected.
    /// </summary>
    public Action? OnOverflow;

    public VideoSocket(Config config)
    {
        _config = config;
    }

    public void Connect()
    {
        _ = ConnectAsync();
    }

    private async Task ConnectAsync()
    {
        var builder = new UriBuilder(_config.Endpoint);
        var query = new StringBuilder(builder.Query.TrimStart('?'));
        void AddParam(string k, string v)
        {
            if (query.Length > 0) query.Append('&');
            query.Append(Uri.EscapeDataString(k)).Append('=').Append(Uri.EscapeDataString(v));
        }
        AddParam("authorization", $"Bearer {_config.AccessToken}");
        if (!string.IsNullOrEmpty(_config.IdToken)) AddParam("id_token", _config.IdToken);
        AddParam("refresh_token", "");
        builder.Query = query.ToString();
        var url = builder.Uri;

        var ws = new ClientWebSocket();
        try
        {
            ws.Options.SetRequestHeader("authorization", $"Bearer {_config.AccessToken}");
            if (!string.IsNullOrEmpty(_config.IdToken)) ws.Options.SetRequestHeader("id_token", _config.IdToken);
            ws.Options.SetRequestHeader("refresh_token", "");
        }
        catch { /* some headers are restricted; query params carry the auth anyway */ }

        _ws = ws;
        _cts = new CancellationTokenSource();
        try
        {
            await ws.ConnectAsync(url, _cts.Token);
        }
        catch (Exception err)
        {
            if (_intentionalClose) return;
            HandleTaskError(err);
            return;
        }

        lock (_sendLock) { _isOpen = true; }
        _everOpened = true;
        _reconnectAttempt = 0;
        _handshakeFailures = 0;
        Console.WriteLine("✓ video WebSocket open");
        SendStartVideo();
        FlushPending();
        _ = ReceiveLoop(ws, _cts.Token);
    }

    /// <summary>fMP4 segment → binary WS frame (buffered while the socket is down).</summary>
    public void SendSegment(byte[] data)
    {
        lock (_sendLock)
        {
            if (_overflowed) return;
            if (!_isOpen || _ws is not { State: WebSocketState.Open })
            {
                _pending.Add(data);
                _pendingBytes += data.Length;
                if (_pendingBytes > MaxPendingBytes)
                {
                    _overflowed = true;
                    _pending.Clear();
                    _pendingBytes = 0;
                    Console.Error.WriteLine(
                        "✗ video buffer overflow during outage — stopping video (audio unaffected)");
                    OnOverflow?.Invoke();
                }
                return;
            }
            var ws = _ws;
            var ct = _cts?.Token ?? CancellationToken.None;
            _ = ws.SendAsync(new ArraySegment<byte>(data), WebSocketMessageType.Binary, true, ct)
                .ContinueWith(t =>
                {
                    if (t.Exception != null && !_intentionalClose)
                        Console.Error.WriteLine($"video send error: {t.Exception.GetBaseException().Message}");
                }, TaskContinuationOptions.OnlyOnFaulted);
        }
    }

    public void SendEnd()
    {
        var meta = new Dictionary<string, object>
        {
            ["callId"] = _config.CallId,
            ["callEvent"] = "END_VIDEO",
        };
        SendJson(meta, "END_VIDEO");
    }

    public void BeginClose()
    {
        _intentionalClose = true;
    }

    public void Close()
    {
        _intentionalClose = true;
        try
        {
            var ws = _ws;
            if (ws is { State: WebSocketState.Open })
                _ = ws.CloseAsync(WebSocketCloseStatus.NormalClosure, null, CancellationToken.None);
        }
        catch { /* ignore */ }
        try { _cts?.Cancel(); } catch { }
        lock (_sendLock) { _isOpen = false; }
    }

    private void SendStartVideo()
    {
        var meta = new Dictionary<string, object>
        {
            ["callId"] = _config.CallId,
            ["callEvent"] = "START_VIDEO",
            ["videoTimeOffsetMs"] = VideoTimeOffsetMs,
        };
        // Same encoder session continuing over a new socket: the server appends
        // to the current segment instead of rotating files.
        if (_startedOnce) meta["videoResume"] = true;
        SendJson(meta, _startedOnce ? "START_VIDEO (resume)" : "START_VIDEO");
        _startedOnce = true;
    }

    private void FlushPending()
    {
        lock (_sendLock)
        {
            if (!_isOpen || _pending.Count == 0 || _ws is not { State: WebSocketState.Open }) return;
            int count = _pending.Count;
            var ws = _ws;
            var ct = _cts?.Token ?? CancellationToken.None;
            foreach (var d in _pending)
                _ = ws.SendAsync(new ArraySegment<byte>(d), WebSocketMessageType.Binary, true, ct);
            _pending.Clear();
            _pendingBytes = 0;
            Console.WriteLine($"↺ flushed {count} buffered video segments after reconnect");
        }
    }

    private void ScheduleReconnect(string why)
    {
        lock (_reconnectLock)
        {
            if (_intentionalClose || _reconnectScheduled || _overflowed) return;
            _reconnectScheduled = true;
            lock (_sendLock) { _isOpen = false; }
            _reconnectAttempt++;
            double delay = Math.Min(MaxBackoff, Math.Pow(2.0, _reconnectAttempt - 1) * 0.5);
            Console.WriteLine($"⟳ video WS reconnect #{_reconnectAttempt} in {delay:F1}s ({why})");
            _ = Task.Delay(TimeSpan.FromSeconds(delay)).ContinueWith(_ =>
            {
                if (_intentionalClose) return;
                lock (_reconnectLock) { _reconnectScheduled = false; }
                Connect();
            });
        }
    }

    private void SendJson(Dictionary<string, object> obj, string label)
    {
        var ws = _ws;
        if (ws is not { State: WebSocketState.Open }) return;
        var json = JsonSerializer.Serialize(obj);
        var bytes = Encoding.UTF8.GetBytes(json);
        var ct = _cts?.Token ?? CancellationToken.None;
        _ = ws.SendAsync(new ArraySegment<byte>(bytes), WebSocketMessageType.Text, true, ct)
            .ContinueWith(t =>
            {
                if (t.Exception != null)
                    Console.Error.WriteLine($"{label} send error: {t.Exception.GetBaseException().Message}");
                else
                    Console.WriteLine($"→ sent {label}");
            });
    }

    private async Task ReceiveLoop(ClientWebSocket ws, CancellationToken ct)
    {
        var buffer = new byte[8192];
        try
        {
            while (ws.State == WebSocketState.Open && !ct.IsCancellationRequested)
            {
                var result = await ws.ReceiveAsync(new ArraySegment<byte>(buffer), ct);
                if (result.MessageType == WebSocketMessageType.Close)
                {
                    lock (_sendLock) { _isOpen = false; }
                    if (_intentionalClose) return;
                    ScheduleReconnect($"server closed {(int?)ws.CloseStatus}");
                    return;
                }
                // The transcriber doesn't send app-level frames back on the
                // video socket; ignore anything received.
            }
            if (!_intentionalClose)
            {
                lock (_sendLock) { _isOpen = false; }
                ScheduleReconnect("receive ended");
            }
        }
        catch (Exception err)
        {
            if (_intentionalClose) return;
            Console.Error.WriteLine($"video WS receive/closed: {err.Message}");
            lock (_sendLock) { _isOpen = false; }
            ScheduleReconnect("receive error");
        }
    }

    private void HandleTaskError(Exception error)
    {
        lock (_sendLock) { _isOpen = false; }
        if (_intentionalClose) return;
        if (!_everOpened)
        {
            _handshakeFailures++;
            if (_handshakeFailures >= MaxHandshakeFailures)
            {
                // Unlike the audio socket, a dead video connection is not fatal
                // to the app: give up on video only (auth failures also hit the
                // audio socket, which owns the loud failure path).
                Console.Error.WriteLine(
                    $"✗ video WebSocket handshake keeps failing — video disabled for this call: {error.Message}");
                _intentionalClose = true;
                OnOverflow?.Invoke();
                return;
            }
        }
        ScheduleReconnect("connect error");
    }
}

File: lma-desktop-capture-app-stack/source/macos/Sources/LMACaptureClient/VideoSocket.swift
import Foundation

/// Second WebSocket, dedicated to the optional desktop-video stream:
///   1. connect wss://.../api/v1/ws  (same endpoint + query-param auth as the
///      audio socket — see the AUTH NOTE in TranscriberSocket.swift)
///   2. send one JSON text frame { callEvent: "START_VIDEO", callId, ... }
///   3. stream fragmented-MP4 segments as binary frames
///   4. send { callEvent: "END_VIDEO" } and close
///
/// Keeping video on its own socket means video bytes can never delay the
/// real-time audio PCM (no head-of-line blocking), and old servers — which
/// ignore the unknown START_VIDEO event and drop binary frames on a session-
/// less socket — degrade to exactly the pre-video behavior.
///
/// Reconnect semantics differ from audio: the fMP4 encoder session on our side
/// keeps running across a socket drop, so the bytes CONTINUE one stream. On
/// reconnect we re-send START_VIDEO with videoResume=true and flush everything
/// buffered — segments must never be dropped mid-stream (unlike audio, a gap
/// makes the remainder undecodable). If the outage outlasts the buffer cap we
/// give up on video for this call (audio is the priority) via onOverflow.
final class VideoSocket: NSObject, URLSessionWebSocketDelegate, URLSessionTaskDelegate {
    private let config: Config
    private var session: URLSession!
    private var task: URLSessionWebSocketTask!
    private var isOpen = false
    private let queue = DispatchQueue(label: "lma.ws.video.send")

    private var intentionalClose = false
    private var reconnectAttempt = 0
    private let maxBackoff: TimeInterval = 10
    private let reconnectQueue = DispatchQueue(label: "lma.ws.video.reconnect")
    private var reconnectScheduled = false
    private var everOpened = false
    private var handshakeFailures = 0
    private let maxHandshakeFailures = 4

    /// True after the first successful START_VIDEO: reconnects resume.
    private var startedOnce = false

    /// ms between the audio stream's start and the first video frame; sent on
    /// START_VIDEO so the server can align video with the transcript timeline.
    var videoTimeOffsetMs: Int = 0

    // Segments buffered while the socket is down. Generous cap (~2 minutes at
    // the configured bitrate) because dropping mid-stream corrupts the video.
    private var pending: [Data] = []
    private var pendingBytes = 0
    private let maxPendingBytes = 32 * 1024 * 1024
    private var overflowed = false

    /// Fired once if the buffer cap is exceeded during an outage — the caller
    /// should stop video capture for this call (audio continues unaffected).
    var onOverflow: (() -> Void)?

    init(config: Config) {
        self.config = config
        super.init()
    }

    func connect() {
        guard var comps = URLComponents(string: config.endpoint) else {
            FileHandle.standardError.write("Bad endpoint URL: \(config.endpoint)\n".data(using: .utf8)!)
            return
        }
        var items = comps.queryItems ?? []
        items.append(URLQueryItem(name: "authorization", value: "Bearer \(config.accessToken)"))
        if !config.idToken.isEmpty {
            items.append(URLQueryItem(name: "id_token", value: config.idToken))
        }
        items.append(URLQueryItem(name: "refresh_token", value: ""))
        comps.queryItems = items
        guard let url = comps.url else { return }
        var req = URLRequest(url: url)
        req.setValue("Bearer \(config.accessToken)", forHTTPHeaderField: "authorization")
        if !config.idToken.isEmpty {
            req.setValue(config.idToken, forHTTPHeaderField: "id_token")
        }
        req.setValue("", forHTTPHeaderField: "refresh_token")

        session = URLSession(configuration: .default, delegate: self, delegateQueue: nil)
        task = session.webSocketTask(with: req)
        task.resume()
        receiveLoop()
    }

    /// fMP4 segment → binary WS frame (buffered while the socket is down).
    func sendSegment(_ data: Data) {
        queue.async {
            guard !self.overflowed else { return }
            guard self.isOpen else {
                self.pending.append(data)
                self.pendingBytes += data.count
                if self.pendingBytes > self.maxPendingBytes {
                    // Outage outlasted the buffer: video for this call is lost
                    // from here. Let the controller stop the capture cleanly.
                    self.overflowed = true
                    self.pending.removeAll(); self.pendingBytes = 0
                    FileHandle.standardError.write(
                        "✗ video buffer overflow during outage — stopping video (audio unaffected)\n".data(using: .utf8)!)
                    DispatchQueue.main.async { self.onOverflow?() }
                }
                return
            }
            self.task.send(.data(data)) { err in
                if let err = err { FileHandle.standardError.write("video send error: \(err)\n".data(using: .utf8)!) }
            }
        }
    }

    func sendEnd() {
        let meta: [String: Any] = [
            "callId": config.callId,
            "callEvent": "END_VIDEO",
        ]
        sendJSON(meta, label: "END_VIDEO")
    }

    func beginClose() { intentionalClose = true }

    func close() {
        intentionalClose = true
        task?.cancel(with: .goingAway, reason: nil)
    }

    private func sendStartVideo() {
        var meta: [String: Any] = [
            "callId": config.callId,
            "callEvent": "START_VIDEO",
            "videoTimeOffsetMs": videoTimeOffsetMs,
        ]
        if startedOnce {
            // Same encoder session continuing over a new socket: the server
            // appends to the current segment instead of rotating files.
            meta["videoResume"] = true
        }
        sendJSON(meta, label: startedOnce ? "START_VIDEO (resume)" : "START_VIDEO")
        startedOnce = true
    }

    private func flushPending() {
        queue.async {
            guard self.isOpen, !self.pending.isEmpty else { return }
            let count = self.pending.count
            for d in self.pending {
                self.task.send(.data(d)) { _ in }
            }
            self.pending.removeAll(); self.pendingBytes = 0
            print("↺ flushed \(count) buffered video segments after reconnect")
        }
    }

    private func scheduleReconnect(_ why: String) {
        reconnectQueue.async { [weak self] in
            guard let self = self else { return }
            if self.intentionalClose || self.reconnectScheduled || self.overflowed { return }
            self.reconnectScheduled = true
            self.isOpen = false
            self.reconnectAttempt += 1
            let delay = min(self.maxBackoff, pow(2.0, Double(self.reconnectAttempt - 1)) * 0.5)
            print("⟳ video WS reconnect #\(self.reconnectAttempt) in \(String(format: "%.1f", delay))s (\(why))")
            self.reconnectQueue.asyncAfter(deadline: .now() + delay) {
                if self.intentionalClose { return }
                self.reconnectScheduled = false
                self.connect()
            }
        }
    }

    private func sendJSON(_ obj: [String: Any], label: String) {
        guard let json = try? JSONSerialization.data(withJSONObject: obj),
              let str = String(data: json, encoding: .utf8) else { return }
        task.send(.string(str)) { err in
            if let err = err {
                FileHandle.standardError.write("\(label) send error: \(err)\n".data(using: .utf8)!)
            } else {
                print("→ sent \(label)")
            }
        }
    }

    private func receiveLoop() {
        task.receive { [weak self] result in
            guard let self = self else { return }
            switch result {
            case .failure(let err):
                if self.intentionalClose { return }
                FileHandle.standardError.write("video WS receive/closed: \(err)\n".data(using: .utf8)!)
                self.scheduleReconnect("receive error")
            case .success:
                self.receiveLoop()
            }
        }
    }

    // MARK: URLSessionWebSocketDelegate

    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask,
                    didOpenWithProtocol protocol: String?) {
        queue.async {
            self.isOpen = true
        }
        everOpened = true
        reconnectAttempt = 0
        handshakeFailures = 0
        print("✓ video WebSocket open")
        sendStartVideo()
        flushPending()
    }

    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask,
                    didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {
        queue.async { self.isOpen = false }
        if intentionalClose { return }
        scheduleReconnect("server closed \(closeCode.rawValue)")
    }

    func urlSession(_ session: URLSession, task: URLSessionTask,
                    didCompleteWithError error: Error?) {
        queue.async { self.isOpen = false }
        guard !intentionalClose else { return }
        if !everOpened {
            handshakeFailures += 1
            if handshakeFailures >= maxHandshakeFailures {
                // Unlike the audio socket, a dead video connection is not fatal
                // to the app: give up on video only (likely an old server that
                // rejects nothing — auth failures would also hit the audio
                // socket, which owns the loud failure path).
                FileHandle.standardError.write(
                    "✗ video WebSocket handshake keeps failing — video disabled for this call\n".data(using: .utf8)!)
                intentionalClose = true
                DispatchQueue.main.async { self.onOverflow?() }
                return
            }
        }
        scheduleReconnect("task error")
    }
}
